cirq/ops/parallel_gate_operation.py

```diff
@@ -0,0 +1,233 @@
+
+            if result is args.available_buffer:
+                args.available_buffer = state
+
+            state = result
+
+        return state
+
+    def _has_unitary_(self) -> bool:
+        return protocols.has_unitary(self.gate)
+
+    def _unitary_(self) -> Union[np.ndarray, NotImplementedType]:
+        # Obtain the unitary for the single qubit gate
+        single_unitary = protocols.unitary(self.gate, NotImplemented)
+
+        # Make sure we actually have a matrix
+        if single_unitary is NotImplemented:
+            return single_unitary
+
+        # Create a unitary which corresponds to applying the single qubit
+        # unitary to each qubit. This will blow up memory fast.
+        unitary = single_unitary
+        for _ in range(len(self.qubits) - 1):
+            unitary = np.outer(unitary, single_unitary)
+
+        return unitary
+
+    def _is_parameterized_(self) -> bool:
+        return protocols.is_parameterized(self.gate)
+
+    def _resolve_parameters_(self, resolver):
+        resolved_gate = protocols.resolve_parameters(self.gate, resolver)
+        return self.with_gate(resolved_gate)
+
+    def _formatted_exponent(self, info: protocols.CircuitDiagramInfo,
+                            args: protocols.CircuitDiagramInfoArgs
+                            ) -> Optional[str]:
+        if info.exponent == 0:
+            return '0'
+
+        # 1 is not shown.
+        if info.exponent == 1:
+            return None
+
+        # Round -1.0 into -1.
+        if info.exponent == -1:
+            return '-1'
+
+        # If it's a float, show the desired precision.
+        if isinstance(info.exponent, float):
+            if args.precision is not None:
+                # funky behavior of fraction, cast to str in constructor helps.
+                approx_frac = Fraction(info.exponent).limit_denominator(16)
+                if approx_frac.denominator not in [2, 4, 5, 10]:
+                    if abs(float(approx_frac)
+                           - info.exponent) < 10 ** -args.precision:
+                        return '({})'.format(approx_frac)
+
+                return '{{:.{}}}'.format(args.precision).format(info.exponent)
+        return repr(info.exponent)
+
+    def _circuit_diagram_info_(self,
+                               args: protocols.CircuitDiagramInfoArgs
+                               ) -> protocols.CircuitDiagramInfo:
+        diagram_info = protocols.circuit_diagram_info(self.gate,
+                                                      args,
+                                                      NotImplemented)
+        if diagram_info == NotImplemented:
+            return diagram_info
+
+        # Extract the formatted exponent and place it in wire symbols so that
+        # every gate on the diagram has an exponent instead of the last one
+        symbol = diagram_info.wire_symbols[0]
+        exponent = self._formatted_exponent(diagram_info, args)
```

Comment:
Ah I see, you're trying to put the exponent on every one of the gates.

Instead of duplicating so much logic between here and `_circuit.py`, we should move this utility code into `CircuitDiagramInfo`. For example, it could have a `rendered_symbols` method.

Reply:
This exponent treatment was only necessary due to the fact that the `_draw_moment_in_diagram` method in `circuit.py` tacks on the exponent to the last symbol in the operation, so it seems like adjusting this behavior is the most straightforward approach. It feels natural to me that if the `connected` keyword argument of a `CircuitDiagramInfo` object is set to `False`, then the exponent should go on every symbol in the operation. An additional explicit keyword argument could be added to `CircuitDiagramInfo` instead. This circumvents the need for the exponent hack here. Does this sound acceptable?

Reply:
That change sounds good to me.